Refactoring URLSession for multiple native protocols

saiHemak opened the pull request:
The code in this PR is a part of #1122 and has been separated to reduce the size of the code change. 

#968 refactored URLSession to decouple it from the HTTP implementation and allow native protocols to be implemented as URLProtocols, uniform with custom protocols. With this PR, we go a step further, and factor out common code which can be used with multiple native protocols. All of this code is moved to a new class called `_NativeProtocol`.

The existing HTTP implementation and the FTP Implementation [#1122] will be reusing `_NativeProtocol`.

saiHemak commented:
@pushkarnk  please review

weissi commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,715 @@
+
+class _NativeProtocol: URLProtocol, _EasyHandleDelegate {
+    internal var easyHandle: _EasyHandle!
+    internal var totalDownloaded = 0
```
can't see this being used anywhere, needed?

saiHemak replied:
I believe ```totalDownloaded``` can be removed .I  had  moved it from  HTTPURLProtocol

weissi commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,715 @@
+}()
+
+class _NativeProtocol: URLProtocol, _EasyHandleDelegate {
+    internal var easyHandle: _EasyHandle!
```
this doesn't seem to get modified, `let`?

saiHemak replied:
```easyHandle ``` is getting initialized during multiple  initis ```init``` and  hence the compiler complains   ``` Immutable value 'self.easyHandle' may only be initialized once ```

weissi commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,715 @@
+    internal var easyHandle: _EasyHandle!
+    internal var totalDownloaded = 0
+    internal lazy var tempFileURL: URL = {
+        let fileName = NSTemporaryDirectory() + NSUUID().uuidString + ".tmp"
```
- don't we have a better/safer way to create temp files somewhere
 - is `NSTemporaryDirectory()` guaranteed to end in a `/` ?

saiHemak replied:
This change has been incorporated into the code base through https://github.com/apple/swift-corelibs-foundation/pull/1195.Hence I have just refactored and pulled it to NativeProtocol as it can be used by multiple protocols like FTP and HTTP

saiHemak commented:
@weissi  -Thanks for the review. Yes its just refactoring of URLSession code .Earlier the code was tightly coupled with HTTP Protocol.To implement the support for other protocols like FTP I moved out the common code in to Native protocol .Please refer to the #1122 for more details on how this code will be used to support FTP and HTTP protocols.

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

ianpartridge commented:
I know this is an inherited problem and not caused by this PR, but the URLSession code is littered with `fatalError()` calls which I think would be better replaced by `assert()` or `preconditionFailure()`. What do you think @pushkarnk?

ianpartridge commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -189,16 +163,78 @@ fileprivate extension _HTTPURLProtocol {
+    /// What action to take
+    override func completionAction(forCompletedRequest request: URLRequest, response: URLResponse) -> _CompletionAction {
+        // Redirect:
+        let httpURLResponse = response as!  HTTPURLResponse
```
The pattern used elsewhere in the codebase is

```swift
guard let response = response as? HTTPURLResponse else {
    fatalError("Reponse was not HTTPURLResponse")
}
```

ianpartridge commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -189,16 +163,78 @@ fileprivate extension _HTTPURLProtocol {
+        //TODO: Should keep track of the number of redirects that this
+        // request has gone through and err out once it's too large, i.e.
+        // call into `failWith(errorCode: )` with NSURLErrorHTTPTooManyRedirects
+        guard case .transferCompleted(response: let response, bodyDataDrain: let bodyDataDrain) = self.internalState else {
```
`response` and `bodyDataDrain` don't seem to be used?

saiHemak replied:
Its been used  .

alblue commented:
Yeah, after perusing #1122 I have the same concern :-) I think it's a good time to bring it up to date with coding practices.

ianpartridge commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,714 @@
+        let url: URL
+        /// Raw headers received.
+        let parsedResponseHeader: _ParsedResponseHeader
+        /// Once the headers is complete, this will contain the response
```
`is` -> `are`

ianpartridge commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,714 @@
+        let requestBodySource: _BodySource?
+        /// Body data received
+        let bodyDataDrain: _NativeProtocol._DataDrain
+        /// Describes what to do with received body data for this transfer:
```
What is this comment for?

ianpartridge commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,714 @@
+        /// The transfer completed.
+        ///
+        /// The easy handle has been removed from the multi handle. This does
+        /// not (necessarily mean the task completed. A task that gets
```
Remove `(`

ianpartridge commented:
I've done a first pass but feel like I only scratched the surface of this PR. There is so much that can be improved - not a criticism of @saiHemak who is working valiantly!

pushkarnk commented:
@ianpartridge Yes, there's a lot of cleanup needed, with the fatalError() calls and in general. However, I am not sure if we should add that here. I'd prefer that done in a different (set of) pull request(s). What do you think?

ianpartridge commented:
Yes, we should split up the work as much as possible.

johnno1962 commented:
Hi, would it be possible to avoid merging this refactoring before https://github.com/apple/swift-corelibs-foundation/pull/1198? Merging in the opposite order will involve a good deal more “conflicts”.

saiHemak commented:
@ianpartridge  @pushkarnk please review ..

pushkarnk commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,725 @@
+
+    fileprivate func notifyDelegate(aboutReceivedData data: Data) {
+        guard let t = self.task else {
+            fatalError("Cannot notify") }
```
The `}` needs to be indented right.

pushkarnk commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,725 @@
+
+    func fill(writeBuffer buffer: UnsafeMutableBufferPointer<Int8>) -> _EasyHandle._WriteBufferResult {
+        guard case .transferInProgress(let ts) = internalState else {
+            fatalError("Requested to fill write buffer, but transfer isn't in progress.") }
```
I think we need to have the `}` on the next line

pushkarnk commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,725 @@
+    ///     `ParsedResponseHeader` with the given line added.
+    func byAppending(headerLine data: Data, headerCompleted: (String) -> Bool) -> _NativeProtocol._ParsedResponseHeader? {
+        // The buffer must end in CRLF
+        guard
```
Can we confirm this is the desired indentation style?

pushkarnk commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -189,16 +163,78 @@ fileprivate extension _HTTPURLProtocol {
+            // before we call delegate API
+            
+            self.internalState = .waitingForRedirectCompletionHandler(response: response, bodyDataDrain: bodyDataDrain)
+            // We need this ugly cast in order to be able to support `URLSessionTask.init()`
```
Is this comment relevant any more?

saiHemak replied:
Though I felt it as irrelevant, I have just retained the old comment as I am not sure of the reason behind this comment.

pushkarnk commented:
Just running a test

@swift-ci please test

pushkarnk commented:
@swift-ci please test

ianpartridge commented on Foundation/URLSession/BodySource.swift:
```diff
@@ -128,12 +128,12 @@ internal final class _HTTPBodyFileSource {
         self.channel.setLimit(highWater: CFURLSessionMaxWriteSize)
     }
-
+    
```
Why does this add extraneous whitespace? Have you set up Xcode correctly?

ianpartridge commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,748 @@
+            return
+        }
+
+        if let response = task?.response {
```
Why not just `ts.response = task?.response` (and change `let ts` to `var ts` above).

ianpartridge commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -365,392 +360,9 @@ fileprivate extension _HTTPURLProtocol._Body {
-    }
-
-    func seekInputStream(to position: UInt64) throws {
-        // We will reset the body sourse and seek forward.
```
It's a shame to lose this comment.

saiHemak replied:
I  have moved these methods to NativeProtocol.swift and updated the comments accordingly

ianpartridge commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -365,392 +360,9 @@ fileprivate extension _HTTPURLProtocol._Body {
-    }
- 
-    func updateProgressMeter(with propgress: _EasyHandle._Progress) {
-        //TODO: Update progress. Note that a single URLSessionTask might
```
Likewise this comment is missing in the moved code. Please retain existing comments.

saiHemak replied:
Same as above

alblue commented:
Can this pull request be rebased (via `git pull --rebase`) so that we can get rid of outstanding conflicts?

saiHemak commented:
@alblue , resolved all the conflicts .

saiHemak commented:
@ianpartridge  @alblue  @pushkarnk , Thanks for the review.Is this PR can be merged now ? PR #1122 is blocked on this

saiHemak commented:
@pushkarnk  can you please review ?

xwu commented on Foundation/URLSession/BodySource.swift:
```diff
@@ -128,7 +128,7 @@ internal final class _HTTPBodyFileSource {
                                        oflag: O_RDONLY, mode: 0, queue: workQueue,
                                        cleanupHandler: {_ in }) else {
-            fatalError("Cant create DispatchIO channel")
+                fatalError("Cant create DispatchIO channel")
```
Please restore the original, correct indent.

xwu commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -10,41 +10,29 @@
+        super.init(task: task, cachedResponse: cachedResponse, client: client)
     }
-
+    
```
Please remove these extraneous changes in indentation.

xwu commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,756 @@
+    }
+}
+
+
```
Please delete extraneous newlines.

xwu commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,756 @@
+}
+
+extension _NativeProtocol {
+    /// An HTTP header being parsed.
```
Is this still always an _HTTP_ header to be parsed?

xwu commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,756 @@
+}
+
+extension _NativeProtocol._ResponseHeaderLines {
+    
```
Please remove this extraneous line.

xwu commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,756 @@
+extension _NativeProtocol._ResponseHeaderLines {
+    
+    func createURLResponse(for URL: URL, contentLength: Int64) -> URLResponse? {
+        return URLResponse(url: URL, mimeType: nil, expectedContentLength: Int(contentLength),textEncodingName: nil)
```
Nit: space between comma and `textEncodingName`.

xwu commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,756 @@
+    /// that ending.
+    /// - Returns: Returning nil indicates failure. Otherwise returns a new
+    ///     `ParsedResponseHeader` with the given line added.
+    func byAppending(headerLine data: Data, headerCompleted: (String) -> Bool) -> _NativeProtocol._ParsedResponseHeader? {
```
I do believe the Swift way of spelling the second parameter would be `isHeaderComplete`. "Completed" is not the right word--it means something different.

xwu commented on Foundation/URLSession/NativeProtocol.swift:
```diff
@@ -0,0 +1,756 @@
+        return byAppending(headerLine: line, headerCompleted: headerCompleted)
+    }
+    
+    private func byAppending(headerLine line: String, headerCompleted: (String) -> Bool) -> _NativeProtocol._ParsedResponseHeader {
```
Likewise here about `headerCompleted`.

xwu commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -804,67 +418,9 @@ internal extension _HTTPURLProtocol {
-    }
+    
+    
+    
```
Please remove multiple consecutive empty lines.

xwu commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -927,12 +483,15 @@ fileprivate extension HTTPURLResponse {
     }
 }
+
+
```
Please remove extra empty lines.

xwu commented on build.py:
```diff
@@ -424,18 +424,18 @@
-	'Foundation/URLSession/http/TransferState.swift',
-	'Foundation/URLSession/http/libcurlHelpers.swift',
-    'Foundation/URLSession/http/HTTPURLProtocol.swift',
+        'Foundation/URLSession/http/HTTPURLProtocol.swift',
```
Almost...indent is too far in by one space :)

xwu commented on Foundation/URLSession/http/HTTPURLProtocol.swift:
```diff
@@ -368,395 +368,9 @@ fileprivate extension _HTTPURLProtocol._Body {
-internal extension _HTTPURLProtocol {
-    /// The data drain.
-    ///
-    /// This depends on what the delegate / completion handler need.
```
This doc comment is missing in the corresponding new version.

saiHemak replied:
Addressed the comments

saiHemak commented:
@xwu , Thanks for your time in reviewing this .I agree that this would take considerable amount of time to review as it cannot be directly compared with old diffs. However, as explained in the description this PR is all about refactoring the existing code  to support multiple native protocols like FTP,HTTP ... Hence most of the code is same and it is just a movement from one file to other file. Refactoring is needed because FTP also reuses the most of the  written for HTTP like header parsing,code used for transfer.For more details, please refer to  PR #1122 - Initial  implementation to support FTP  re-uses the same code.

xwu commented:
@saiHemak For reviewers' benefit, can you please document explicitly, then, which code was moved from where, and what code is new? This could be done in text, but it would also be sufficient to have two commits here, one only moving existing code--with no additions whatsoever, maintaining all existing comments and formatting--then another commit with your changes to the code.

saiHemak commented:
@xwu 

Thanks for the inputs ..  

Actually, though the code is  just a movement, few required the renaming too ..
So I am looking for a way on how can I split the commits ..

Here are the details of refactoring  and reason behind  refactoring ..

  1. Moved lib curl specific code [libcurlHelpers.swift,EasyHandle.swift  and  MultiHandle.swift] from  URLSession/http   to URLSession/libcurl 
**Reason :** libcurl specific code  is common and is not specific to HTTP .Hence moved out of the HTTP folder to a common licurl folder  
2.Created  a class NativeProtocol which basically holds the common implementation of native  protocols like HTTP , FTP  and so on .It implements the URLProtocol and EashHandleDelegate .
**Reason:** Need of a common class which  holds the basic functionality and can be extended by respective protocol implementations.

3. Moved  enableLibcurlDebugOutput and enableDebugOutput  debug flags  to  NativeProtocol .These can be enabled on EasyHandle irrespective of the protocol. 
4.Moved _HTTPURLProtocol.InternalState from HTTPURLProtocol  to  _NativeProtocol.InternalState .Moved ```inits``` from HTTPURLProtocol to NativeProtocol
**Reason:**  Inits  are  initializing  common  variables like easyHandle,internal state and temporary file .
 5. Modified HTTPURLProtocol to extend from _NativeProtocol and  inits to just call super.init()   6. Moved easyHandle and tempFileURL instance variables to NativeProtocol  7. Moved implemented methods   _HTTPURLProtocol: _EasyHandleDelegate .
Reason :

EasyHandleDelegate methods  provides functionality for transfer  irrespective of protocol   to NativeProtocol- 

fill(writeBuffer buffer: UnsafeMutableBufferPointer<Int8>) ,
notifyDelegate(aboutUploadedData),
notifyDelegate(aboutReceivedData),
didReceive(data).
transferCompleted(withError error: NSError?)
createTransferState()
createTransferBodyDataDrain

However few methods like below are protocol specific hence have been overridden in respective protocol classes .

 ```didReceive(headerData data: Data, contentLength: Int64) ```
 Though the headers basically  ends with  \r\n  irrespective of protocol the header format in  FTP  is different from HTTP. For e.g.: Headers in FTP looks like a plain text header with status code  and  description unlike the key-value pair headers in HTTP

8.NativeProtocol  has definition for  configureEasyHandle(easyHandle) which needs concrete implementation  which is protocol specific by sub-classes like FTP and HTTP.
**Reason:**
HTTP requires some more additional configurations like adding the headers, setting allowed protocols. 
9.Renamed the HTTPBodySource to BodySource  and moved out of the HTTP folder.
 **Reason:** The code respective to data handling is same irrespective to protocol   10. HTTPMessage._HTTPCharacters  has been moved  and renamed to _NativeProtocol.Delimiters Reason: Refer to next point 

11. Moved  HTTPURLProtocol. _ParsedResponseHeader  from HTTPMessage.swift to NativeProtocol.swift under _NativeProtocol._ParsedResponseHeader .
**Reason:**  This implementation basically does header parsing based on \r\n  delimiters. Except the  fact that empty header is received  in case of HTTP to mark as Header completion and  150 status marks the open of data communication in FTP.  12. Moved and renamed  HTTPURLProtocol._HTTPTransferState from TransferState.swift to _Nativeprotocol.TransferState . Moved related 

**Reason:** Moved  the code to Native Protocol as Transfer states have 1-to-1 with EasyHandle and irrespective of protocol  

As the complete TransferState code has been aligned to Native protocol TransferState.swift  file has been removed .


New Implementation :

NativeProtocol.validateHeaderComplete 
**Reason :**
 Should  not throw error incase of  simple-responseshttps://github.com/apple/swift-corelibs-foundation/pull/1097

isCompleteHeader -> Inner method  
**Reason:**

Refer to point 11 reason.In order to support both protocols this method has been introduced  
-> which returns to true when the header is empty  in case of HTTP.
-> Returns true if the header starts with 150

alblue commented:
It would be good to split the commits, with each commit message outlining the points you have made above, so that anyone reviewing the changes afterwards would have the context without having to go back to a GitHub PR to try and understand the why. It would also help reviewing the changes independently and validate each one separately, including that all of the moves are done without changing code (for sanity) and to then ensure that such moves continue to be built via the PR testing mechanism, instead of just the end result.

saiHemak commented:
@alblue @pushkarnk @xwu  I have split the commit ,please review

alblue commented:
The delete/add commits (64429de and a44a9f9) should really be one commit, not two. That way git will be able to figure out that there has been a move between files, rather than a mass-delete and mass-add. In fact, it might be better to spin that combined commit (of 64429de and a44a9f9) into its own separate review first of all, rebased against master. Once that can be verified it's not breaking any tests on its own, it can be merged - and then the body of this PR can be rebased upon it and we can review the changes subsequently.

So, I recommend:

- cherry pick (and squash) 64429de and a44a9f9 onto master
- create a new PR with that combined commit
- get that reviewed and merged
- revisit the changes on this PR with that base line

Sound good?

saiHemak commented:
@alblue : Sure

xwu commented:
Likewise subsequent pairs of commits that move code. The point is to have git demonstrate that each such a change is *only* moving code, never adding or deleting anything. It is not possible to see this on inspection when one commit removes the code and another adds it back.

alblue commented:
Since we're breaking this apart and filing separate changes, I'm going to close this request as is. Will be happy to look at the incremental changes (such as the one merged previously) in order to get the functionality in. Thanks for keeping up with this!
